File: sdk/dart/lib/vaultkey.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;

/// A Dart SDK client for VaultKey secret manager.
class Vaultkey {
  /// The host URL of the VaultKey server.
  final String host;

  /// The API key used for authenticating with the VaultKey server.
  final String apiKey;

  /// Constructor taking [host] (default: 'http://localhost:8080') and [apiKey].
  Vaultkey({
    String host = 'http://localhost:8080',
    required this.apiKey,
  }) : host = host.endsWith('/') ? host.substring(0, host.length - 1) : host;

  /// Helper method to perform HTTP requests to the VaultKey server.
  Future<dynamic> _request(String path) async {
    final url = Uri.parse('$host$path');
    final headers = {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $apiKey',
    };

    final response = await http.get(url, headers: headers);

    if (response.statusCode < 200 || response.statusCode >= 300) {
      String errorMessage;
      try {
        final data = jsonDecode(response.body);
        errorMessage = data['error'] ?? 'Request failed with status: ${response.statusCode}';
      } catch (_) {
        errorMessage = 'Request failed with status: ${response.statusCode}';
      }
      throw Exception('VaultKey request failed: $errorMessage (status: ${response.statusCode})');
    }

    if (response.body.isEmpty) {
      return null;
    }
    return jsonDecode(response.body);
  }

  /// Retrieves the decrypted value of a single secret.
  Future<String> get(String key, [String project = 'default', String env = 'production']) async {
    final path = '/v1/secrets/${Uri.encodeComponent(key)}?project=${Uri.encodeComponent(project)}&environment=${Uri.encodeComponent(env)}';
    final res = await _request(path);
    if (res is Map && res.containsKey('value')) {
      return res['value'] as String;
    }
    return '';
  }

  /// Lists secret metadata.
  Future<List<dynamic>> list([String project = 'default', String env = 'production']) async {
    final path = '/v1/secrets?project=${Uri.encodeComponent(project)}&environment=${Uri.encodeComponent(env)}';
    final res = await _request(path);
    if (res is List) {
      return res;
    }
    return [];
  }

  /// Fetch all decrypted secrets in the scope as a Map in a single request (using `/v1/secrets/values`).
  Future<Map<String, String>> values([String project = 'default', String env = 'production']) async {
    final path = '/v1/secrets/values?project=${Uri.encodeComponent(project)}&environment=${Uri.encodeComponent(env)}';
    final res = await _request(path);
    if (res is Map) {
      return res.map((key, value) => MapEntry(key.toString(), value.toString()));
    }
    return {};
  }

  /// Fetches the secrets and returns a Map.
  ///
  /// Note: [Platform.environment] is read-only in Dart. Therefore, we cannot
  /// automatically write/inject secrets directly into the system's environment variables.
  /// Instead, this method fetches the secrets and returns them as a Map for in-memory use.
  Future<Map<String, String>> inject([String project = 'default', String env = 'production']) async {
    return values(project, env);
  }
}

File: sdk/python/vaultkey.py
import os
import urllib.parse
import urllib.request
import json

class Vaultkey:
    def __init__(self, api_key: str, host: str = "http://localhost:8080"):
        self.host = host.rstrip("/")
        self.api_key = api_key

    def _request(self, path: str, method: str = "GET", data: dict = None) -> dict:
        url = f"{self.host}{path}"
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.api_key}"
        }
        
        req_data = None
        if data is not None:
            req_data = json.dumps(data).encode("utf-8")
            
        req = urllib.request.Request(url, data=req_data, headers=headers, method=method)
        
        try:
            with urllib.request.urlopen(req) as res:
                body = res.read().decode("utf-8")
                return json.loads(body) if body else {}
        except urllib.error.HTTPError as e:
            err_body = e.read().decode("utf-8")
            try:
                msg = json.loads(err_body).get("error", str(e))
            except Exception:
                msg = str(e)
            raise Exception(f"VaultKey request failed: {msg} (status: {e.code})")

    def get(self, key: str, project: str = "default", env: str = "production") -> str:
        """Retrieves a single decrypted secret value."""
        path = f"/v1/secrets/{urllib.parse.quote(key)}?project={urllib.parse.quote(project)}&environment={urllib.parse.quote(env)}"
        res = self._request(path)
        return res.get("value", "")

    def list(self, project: str = "default", env: str = "production") -> list:
        """Lists active secret keys and metadata."""
        path = f"/v1/secrets?project={urllib.parse.quote(project)}&environment={urllib.parse.quote(env)}"
        return self._request(path)

    def values(self, project: str = "default", env: str = "production") -> dict:
        """Retrieves all decrypted secrets in the scope as a dictionary."""
        path = f"/v1/secrets/values?project={urllib.parse.quote(project)}&environment={urllib.parse.quote(env)}"
        return self._request(path)

    def inject(self, project: str = "default", env: str = "production") -> None:
        """Fetches all active secrets and injects them into os.environ in-memory."""
        secrets = self.values(project, env)
        for k, v in secrets.items():
            os.environ[k] = v
